### PR TITLE
Cleaned up the IndexAwarePredicate API

### PR DESCRIPTION
**Pull request by vojtechtoman**
Removed the references to internal APIs.
Introduced new public APIs: Index, QueryContext,
and QueryableEntry. In the process, merged the
old internal Index interface into InternalIndex
and renamed the original internal QueryableEntry
class to QueryableEntryImpl.

The new public Index interface contains only
"read" methods (retrieving entries from the index).

Fixes https://github.com/hazelcast/hazelcast/issues/13226

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapContainer.java [hunk 1]
```diff
@@ -386,7 +386,7 @@ public String getQuorumName() {
         return recordFactoryConstructor;
     }
 
-    public QueryableEntry newQueryEntry(Data key, Object value) {
+    public QueryableEntryImpl newQueryEntry(Data key, Object value) {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/map/impl/MapContainer.java]: The quaryable entry gives me sleepless nights. But now we have an abstraction for this ugly thing?

File: hazelcast/src/main/java/com/hazelcast/query/QueryContext.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.query;
+
+/**
+ * Provides the context for queries execution.
+ */
+public interface QueryContext {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/query/QueryContext.java]: Does this need to be exposed to the end user?

File: hazelcast/src/main/java/com/hazelcast/query/QueryableEntry.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+package com.hazelcast.query;
+
+import java.util.Map;
+
+/**
+ * Represents a Queryable Entry, which means searched an indexed by SQL query or predicate.
+ * @param <K></K> the key type
+ * @param <V></V> the value type
+ */
+public interface QueryableEntry<K, V> extends Map.Entry<K, V> {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/query/QueryableEntry.java]: Can we come up with a better name for this?

Comment (vojtechtoman) [reply hazelcast/src/main/java/com/hazelcast/query/QueryableEntry.java]: We can probably get rid of this interface altogether and stick with just `Map.Entry` on the public API level. The reason why I introduced this interface was to keep our hands free if we need to add additional methods, extend another interface etc.

**Comment by devozerov** [comment]
This might be a kind of step backward, but speaking of user experience - do we need `IndexAwarePredicate` on public API in the first place? Since this is a task for 4.0 and API is pretty old, it might be a good moment to reconsider it in favor of simplicity.

If trying to draw parallels with SQL, users are typically expect that the engine will decide whether to use index or not automatically. And this is indeed already the case for most of existing predicates. 

Can we imagine a use case when user would decide to implement it's own `IndexAwarePredicate`?

**Comment by devozerov** [comment]
Summary of private discussion: we are considering possibility to hide IndexAwarePredicate because it appears to complex for end users.

**Comment by mmedenjak** [comment]
@vojtechtoman is this PR still valid if we decided that `IndexAwarePredicate` is going to be made internal? Can we close it?

**Comment by vojtechtoman** [comment]
Let's close.

**Pull request closed without merging**